examples/autoignition-mpi.py

```diff
@@ -244,67 +298,144 @@ def main(ctx_factory=cl.create_some_context, casename="autoignition", use_leap=F
                            dt=current_dt, cfl=current_cfl, eos=eos,
                            t_final=t_final, constant_cfl=constant_cfl)
 
+    def my_graceful_exit(cv, step, t, do_viz=False, do_restart=False, message=None):
+        if rank == 0:
+            logger.info("Errors detected; attempting graceful exit.")
+        if do_viz:
+            my_write_viz(cv, step, t)
+        if do_restart:
+            my_write_restart(state=cv, step=step, t=t)
+        if message is None:
+            message = "Fatal simulation errors detected."
+        raise RuntimeError(message)
+
+    def my_write_viz(cv, step, t, dv=None, production_rates=None):
+        viz_fields = [("cv", cv)]
+        if dv is not None:
+            viz_fields.append(("dv", dv))
+        if production_rates is not None:
+            viz_fields.append(("production_rates", production_rates))
+        write_visfile(discr, viz_fields, visualizer, vizname=casename,
+                      step=step, t=t, overwrite=True)
+
+    def my_write_restart(state, step, t):
+        rst_fname = rst_pattern.format(cname=casename, step=step, rank=rank)
+        rst_data = {
+            "local_mesh": local_mesh,
+            "state": state,
+            "t": t,
+            "step": step,
+            "global_nelements": global_nelements,
+            "num_parts": nproc
+        }
+        from mirgecom.restart import write_restart_file
+        write_restart_file(actx, rst_data, rst_fname, comm)
+
+    def my_health_check(dv, dt):
+        health_error = False
+        from mirgecom.simutil import check_naninf_local, check_range_local
+        if check_naninf_local(discr, "vol", dv.pressure) \
+           or check_range_local(discr, "vol", dv.pressure, 1e5, 2.4e5):
+            health_error = True
+            logger.info(f"{rank=}: Invalid pressure data found.")
+
+        if check_range_local(discr, "vol", dv.temperature, 1.4e3, 3.3e3):
+            health_error = True
+            logger.info(f"{rank=}: Invalid temperature data found.")
+
+        if dt < 0:
+            health_error = True
+            if rank == 0:
+                logger.info("Global DT is negative!")
```

Comment:
Just a clarifying comment: when and how would `dt` ever go negative?

Reply:
Not sure it ever can, although I do not know the inner workings of `characteristic_lengthscales`.  Can it be negative?  The other thing is wavespeeds, and by construction this cannot be negative. This is likely a superfluous check. If gamma, P, or rho are neg, then speed-of-sound will be complex.  But previous health check would have already caught the issue by then.

Reply:
@MTCam `characteristic_lengthscales` should _never_ be negative. It works with areas and distances which can not be negative

Reply:
I don't think DT can ever be negative then, unless the user prescribes one.  Are we confident enough about that to take this check out?  I would just check speed-of-sound but calculating that is not super cheap and it has already been calculated for use in DT.